File: src/ethernity/encoding/chunking.py
#!/usr/bin/env python3
# Copyright (C) 2026 Alex Stoyanov
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along with this program.
# If not, see <https://www.gnu.org/licenses/>.

from __future__ import annotations

from typing import Iterable, Sequence

from .framing import DOC_ID_LEN, VERSION, Frame, decode_frame
from .zbase32 import decode_fallback_lines as _decode_fallback_lines

DEFAULT_CHUNK_SIZE = 1024


def chunk_payload(
    payload: bytes,
    *,
    doc_id: bytes,
    frame_type: int,
    chunk_size: int = DEFAULT_CHUNK_SIZE,
    version: int = VERSION,
) -> list[Frame]:
    if not payload:
        raise ValueError("payload cannot be empty")
    if len(doc_id) != DOC_ID_LEN:
        raise ValueError(f"doc_id must be {DOC_ID_LEN} bytes")
    if chunk_size <= 0:
        raise ValueError("chunk_size must be positive")

    total = (len(payload) + chunk_size - 1) // chunk_size
    base_size = len(payload) // total
    remainder = len(payload) % total

    frames: list[Frame] = []
    offset = 0
    for idx in range(total):
        size = base_size + (1 if idx < remainder else 0)
        start = offset
        end = start + size
        frames.append(
            Frame(
                version=version,
                frame_type=frame_type,
                doc_id=doc_id,
                index=idx,
                total=total,
                data=payload[start:end],
            )
        )
        offset = end
    return frames


def reassemble_payload(
    frames: Sequence[Frame],
    *,
    expected_doc_id: bytes | None = None,
    expected_frame_type: int | None = None,
) -> bytes:
    if not frames:
        raise ValueError("no frames provided")

    doc_id = expected_doc_id or frames[0].doc_id
    frame_type = expected_frame_type if expected_frame_type is not None else frames[0].frame_type
    total = frames[0].total
    version = frames[0].version

    if len(doc_id) != DOC_ID_LEN:
        raise ValueError(f"doc_id must be {DOC_ID_LEN} bytes")
    if total <= 0:
        raise ValueError("total must be positive")

    seen: dict[int, Frame] = {}
    for frame in frames:
        if frame.doc_id != doc_id:
            raise ValueError("mismatched doc_id")
        if frame.frame_type != frame_type:
            raise ValueError("mismatched frame_type")
        if frame.total != total:
            raise ValueError("mismatched total")
        if frame.version != version:
            raise ValueError("mismatched version")
        if frame.index < 0:
            raise ValueError("index must be non-negative")
        if frame.index >= total:
            raise ValueError("index must be < total")
        if frame.index in seen:
            existing = seen[frame.index]
            if existing.data != frame.data:
                raise ValueError("conflicting duplicate frames detected")
            continue
        seen[frame.index] = frame

    if len(seen) != total:
        raise ValueError("missing frames")

    return b"".join(seen[idx].data for idx in range(total))


def fallback_lines_to_frame(lines: Iterable[str]) -> Frame:
    data = _decode_fallback_lines(lines)
    return decode_frame(data)

File: kit/app/extensions/chunking.js
/*
 * Copyright (C) 2026 Alex Stoyanov
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with this program.
 * If not, see <https://www.gnu.org/licenses/>.
 */

import { sha256 } from "@noble/hashes/sha2.js";

const ROLLING_HASH_MASK = (1n << 64n) - 1n;
const ROLLING_WINDOW_SIZE = 64;
const MIN_MASK_BITS = 4;
const GEAR_TABLE = buildGearTable();

export function canonicalChunkRefsForBytes(data, chunking) {
  return defaultExtensionChunker(data, chunking).map((chunk) => [sha256(chunk), chunk.length]);
}

export function defaultExtensionChunker(data, profile) {
  if (!data.length) {
    return [];
  }
  const chunks = [];
  let start = 0;
  const [primaryMask, secondaryMask] = rollingMasks(profile.targetSize);
  while (start < data.length) {
    const chunkEnd = nextChunkBoundary(data, {
      start,
      minSize: profile.minSize,
      targetSize: profile.targetSize,
      maxSize: profile.maxSize,
      primaryMask,
      secondaryMask,
    });
    chunks.push(data.slice(start, chunkEnd));
    start = chunkEnd;
  }
  return chunks;
}

function rollingMasks(targetSize) {
  const targetBits = Math.max(MIN_MASK_BITS, Math.round(Math.log2(Math.max(2, targetSize))));
  const primaryBits = Math.min(63, targetBits);
  const secondaryBits = Math.max(MIN_MASK_BITS, targetBits - 2);
  return [(1n << BigInt(primaryBits)) - 1n, (1n << BigInt(secondaryBits)) - 1n];
}

function nextChunkBoundary(data, options) {
  const totalLen = data.length;
  const { start, minSize, targetSize, maxSize, primaryMask, secondaryMask } = options;
  if (totalLen - start <= minSize) {
    return totalLen;
  }
  const minEnd = Math.min(totalLen, start + minSize);
  const targetEnd = Math.min(totalLen, start + targetSize);
  const maxEnd = Math.min(totalLen, start + maxSize);
  let fingerprint = 0n;
  const window = new Uint8Array(ROLLING_WINDOW_SIZE);
  let windowCount = 0;
  let windowPos = 0;
  for (let index = start; index < maxEnd; index += 1) {
    const byteValue = data[index];
    if (windowCount < ROLLING_WINDOW_SIZE) {
      fingerprint = rollFingerprint(fingerprint, byteValue);
      window[windowPos] = byteValue;
      windowPos = (windowPos + 1) % ROLLING_WINDOW_SIZE;
      windowCount += 1;
    } else {
      const outgoing = window[windowPos];
      window[windowPos] = byteValue;
      windowPos = (windowPos + 1) % ROLLING_WINDOW_SIZE;
      fingerprint =
        (rotateLeft(fingerprint, 1) ^ GEAR_TABLE[outgoing] ^ GEAR_TABLE[byteValue]) &
        ROLLING_HASH_MASK;
    }
    if (index + 1 < minEnd) {
      continue;
    }
    const mask = index + 1 < targetEnd ? primaryMask : secondaryMask;
    if ((fingerprint & mask) === 0n) {
      return index + 1;
    }
  }
  return maxEnd;
}

function rollFingerprint(current, byteValue) {
  return rotateLeft(current, 1) ^ GEAR_TABLE[byteValue];
}

function rotateLeft(value, shift) {
  const normalized = BigInt(shift & 63);
  return ((value << normalized) | (value >> (64n - normalized))) & ROLLING_HASH_MASK;
}

function buildGearTable() {
  let state = 0x9e3779b97f4a7c15n;
  const values = [];
  for (let index = 0; index < 256; index += 1) {
    state ^= (state >> 12n) & ROLLING_HASH_MASK;
    state ^= (state << 25n) & ROLLING_HASH_MASK;
    state ^= (state >> 27n) & ROLLING_HASH_MASK;
    state = (state * 0x2545f4914f6cdd1dn) & ROLLING_HASH_MASK;
    values.push(state);
  }
  return values;
}
